ts: Migrate people.js to TypeScript.

Lalit3716 opened the pull request:
This PR follows the work done by @PIG208 in #24653 which is to migrate people.js into TypeScript.
The majority of refactoring required for this migration is already been done.

### Remaining work / current Challenges:

<p>As also mentioned in #24653 we still have some things to figure out:</p>

- [x] Fix some es-lint errors.
- [x] Figuring out correct types for `Messages`.
- [x] `_add_user` currently can also be called with `extract_people_from_message` that adds a user (for deactivated users that aren't originally populated) containing only a subset of required fields on User. The involvement of these objects makes it type-unsafe whenever the global mappings for user storage are accessed.
- [x] In many cases, we do not handle optional values when doing a lookup on one of the global mappings. The current approach would be silencing the typing error by appending ! to assert non-undefinedness without changing the original implementation, but we might want further refactorings to avoid using it.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PIG208 commented on web/src/types.ts:
```diff
@@ -6,19 +6,111 @@ export type MatchedMessage = {
+export type MessageCleanReaction = {
+    class: string;
+    count: number;
+    emoji_alt_code: string;
```
`emoji_alt_code` is a `boolean`:
https://github.com/zulip/zulip/blob/660ccccf681c1036584c96dbe5068acacb5ff82f/web/src/user_status.js#L60-L74

PIG208 commented on web/src/types.ts:
```diff
@@ -6,19 +6,111 @@ export type MatchedMessage = {
+};
 
 // TODO/typescript: Move this to message_store
 export type RawMessage = {
```
We can use this type in the future migration of the `message_helper`, `message_store` family of modules.
Marking the transition from `RawMessage` to `Message` can be tricky, because the mutation might happen in various places. `message_helper.process_new_message` may or may not get a `RawMessage`; the boolean-related fields are added in `message_store.set_message_booleans`; most of the functions in `reactions` calls `reactions.set_clean_reactions` to add `message.clean_reactions`. We need to determine where these transitions might happen to ensure that the types are correct. For this PR, I'm verifying if the functions called with `Message` in `people` indeed get a final `Message` object with all the fields added.

PIG208 commented on web/src/types.ts:
```diff
@@ -6,19 +6,111 @@ export type MatchedMessage = {
+    unread: boolean;
+    url: string;
     topic: string;
+    to_user_ids: string;
```
`to_user_ids` only appears in direct messages

Lalit3716 replied:
Made it optional 👍

PIG208 commented on web/src/types.ts:
```diff
@@ -6,19 +6,111 @@ export type MatchedMessage = {
+};
+
+// TODO/typescript: Move this to message_store
+export type Message = Omit<RawMessage, "reactions"> & {
```
I propose that we can group the following properties by where they are added, not necessarily into different types, at least for now. For example, we can group the booleans added by `set_message_booleans` with a single line comment above mentioning the source.

PIG208 commented:
I was playing around with `Message`, and implemented a zod schema only for reviewing purposes. Despite the issues mentioned in the previous comments, there are some properties that are missing in certain cases. I haven't got the time to look into these in detail yet. You can open your browser console and play around a little bit to exercise the code and see what are some unique `zod` errors you can get.

Some of the errors I have collected so far:
- Missing the boolean properties from the `extract_people_from_message` code path when `initialize_everything` is called
- Missing `url` for `get_pm_emails`
- Missing `url` and `reply_to` for `get_pm_fullnames`

There may be more if you edit messages, add reactions, view huddle messages and etc.

PIG208 commented:
Some testing-only change is created as a PR in your repo so that you can test it out!
https://github.com/Lalit3716/zulip/pull/3

Lalit3716 commented:
Thanks @PIG208 for the detailed review, I will play around with the testing PR and fix types accordingly!

PIG208 commented on web/src/people.ts:
```diff
@@ -130,7 +130,7 @@ export function validate_user_ids(user_ids: number[]): number[] {
 
     if (bad_ids.length > 0) {
-        blueslip.warn(`We have untracked user_ids: ${bad_ids}`);
+        blueslip.warn(`We have untracked user_ids`, {bad_ids});
```
We can use double quotes now that no template strings are needed.

Lalit3716 commented:
@PIG208 Updates:
- If you follow `message_helper.process_new_message` path then we can conclude that `extract_people_from_message` and `maybe_incr_recipient_count` are not exactly `Message` type but rather a more specific `MessageWithBooleans` type so changed the types for these functions accordingly.
- Added some comments in `Message` type to point out sources where the additional properties are added to the raw message.
- Missing `reply_to` for `get_pm_fullnames` was because in `message_helper.process_new_message` while setting `message.reply_to` we call `get_pm_fullnames` and hence it gets `reply_to` as undefined.
```javascript
        case "private":
            message.is_private = true;
            message.reply_to = util.normalize_recipients(message_store.get_pm_emails(message));
```
I fixed this by modifying the types for `people.pm_with_user_ids` -
```typescript
export function pm_with_user_ids(
    message: Message & {reply_to?: string; url?: string},
): number[] | undefined {
```

PIG208 commented:
Thanks for tracking this down!

PIG208 commented:
The summary message for the commit "(squash) people: Type user-related functions."
> `_add_user` currently can also be called with `extract_people_from_message`
that adds a user (for deactivated users that aren't originally populated)
containing only a subset of required fields on `User`. The involvement of
these objects makes it type-unsafe whenever the global mappings for user
storage are accessed.
no longer holds, right? Because we have `make_user` now to ensure that `_add_user` always takes a valid `User`. Eventually, the commit messages for all the to-be-squashed commits will be combined into a single commit when merging.

PIG208 commented on web/src/types.ts:
```diff
@@ -63,17 +73,60 @@ export type RawMessage = {
     stream_id?: number;
     subject: string;
-    submessages: string[];
+    submessages: Submessage[];
```
Nice catch! It turned out that our test cases which have been using `string[]` are not so representative of the real data.

PIG208 commented on web/src/people.ts:
```diff
@@ -14,15 +14,66 @@ import * as settings_config from "./settings_config";
+
+export type User = {
+    user_id: number;
+    delivery_email?: string;
```
`delivery_email` is nullable now due to the feature level 163 change.

PIG208 commented on web/src/people.ts:
```diff
@@ -14,15 +14,66 @@ import * as settings_config from "./settings_config";
+    is_billing_admin: boolean;
+    is_bot: boolean;
+    bot_type?: number;
+    bot_owner_id?: number;
```
Both `bot_type` and `bot_owner_id` are also nullable.

PIG208 replied:
Note that the server has an undocumented behavior that `bot_type` and `bot_owner_id` do not appear when the user is not `null`. We could probably work to make that more accurately documented/behave consistently, but that will be a task of its own.
https://github.com/zulip/zulip/blob/main/zerver/lib/users.py#L501-L507

PIG208 commented on web/src/people.ts:
```diff
@@ -14,15 +14,66 @@ import * as settings_config from "./settings_config";
-let my_user_id;
+export type ProfileData = {
+    value: string;
+    rendered_value: string;
```
`rendered_value` is optional.
https://github.com/zulip/zulip/blob/b5e572811257a33a25494af34a5231823cbe5f02/zerver/openapi/zulip.yaml#L562-L570

Lalit3716 commented:
@PIG208 Updates:
- Refined User types according to your reviews!

PIG208 commented:
> Used non-null assertion operator when retreiving people objects from the global
mappings. We could possibly assert statements here for better type safety?

Agreed. Let's use assertions instead here then. I have no further feedback other than that.

PIG208 commented:
One potential improvement is adding some `zod` schemas parsing the users and messages from the server.
I think we had made quite a lot of progress in this PR. So, it is probably worth splitting that work into a follow-up.

PIG208 commented:
@andersk will you be able to also review this?

PIG208 commented:
@zulipbot add "maintainer review"

PIG208 commented:
@zulipbot add "area: typescript migration"

timabbott commented on web/src/people.js:
```diff
@@ -654,7 +654,7 @@ export function exclude_me_from_string(user_ids_string) {
 
 export function format_small_avatar_url(raw_url) {
-    const url = new URL(raw_url, location);
+    const url = new URL(raw_url, location.origin);
```
This code definitely works, and the stringification is documented: https://developer.mozilla.org/en-US/docs/Web/API/URL/URL

I think it's fine to do this change but the commit message looks to be wrong.

timabbott commented on web/src/people.js:
```diff
@@ -1328,8 +1328,6 @@ export function extract_people_from_message(message) {
             user_id,
             full_name: person.full_name,
-            is_admin: person.is_realm_admin || false,
-            is_bot: person.is_bot || false,
```
I think this should instead be hardcoding these to false. I'm not sure, but I kinda suspect this commit alone breaks the app and it's only fixed when you introduce `make_user` here later; it might make sense to just squash to avoid the intermediate state -- the actual problem here is that this is not a full user object.

timabbott commented on web/src/people.js:
```diff
@@ -1282,6 +1282,30 @@ export function report_late_add(user_id, email) {
 
+function make_user(partial_user) {
+    return {
+        user_id: partial_user.user_id ?? -1,
```
The `user_id`, `email`, and `full_name` parameters need to be required parameters.

timabbott commented on web/src/people.js:
```diff
@@ -1282,6 +1282,30 @@ export function report_late_add(user_id, email) {
+        bot_owner_id: partial_user.bot_owner_id ?? undefined,
+        bot_type: partial_user.bot_type ?? 0,
+        is_moderator: partial_user.is_moderator ?? false,
+    };
```
I think a number of these defaults are dangerous -- for example, `0` is not a valid possible value for `role`. My current thinking is we should rewrite this function to accept just the 3 parameters that we call it with, and pick a consistent set of values that are the defaults for a newly created user, and perhaps an extra new field `is_missing_server_data` that we can use to keep track of the fact that this is a fake user object.

timabbott commented on web/src/people.ts:
```diff
@@ -14,15 +14,67 @@ import * as settings_config from "./settings_config";
+    is_owner: boolean;
+    is_admin: boolean;
+    is_guest: boolean;
+    is_moderator?: boolean;
```
This should be just a boolean, not optional, right?

timabbott commented on web/src/people.ts:
```diff
@@ -14,15 +14,67 @@ import * as settings_config from "./settings_config";
+    bot_owner_id?: number | null;
+    role: number;
+    timezone: string;
+    avatar_url?: string;
```
Should this have a `| null`?

Lalit3716 replied:
Yeah I missed it.

timabbott commented on web/src/people.ts:
```diff
@@ -375,7 +376,7 @@ export function get_user_time(user_id: number): string | undefined {
-
+    assert(user_profile !== undefined, "Unknown person in get_user_type");
     return settings_config.user_role_map.get(user_profile.role);
 }
```
So while we expect our current user ID to always exist, I don't see a reason to expect callers of this to have checked that the user ID in question is a real one. So either that needs to be documented, or this function needs to have different error handling. 

Since this part is an API change, I think it might make sense to split it out to a prep PR.

timabbott commented:
This is progressing well. I posted a batch of comments, mostly on the prep comments before the first big migration commit.

@Lalit3716 can you make a separate PR for those changes? I think some of them need some care to avoid unintended functional changes that might introduce bugs, and we should be able to do faster iteration on a smaller PR that is just those.

Once we have those details resolved and that PR integrated, I expect we'll be pretty close to being able to merge this.

Lalit3716 commented:
@timabbott @PIG208 Sorry for the delay on this, I extracted the prep commits and did the API change which Tim mentioned in #26042 .

amanagr commented:
@Lalit3716 @PIG208 can any of you describe what's the current status of this PR?

PIG208 commented:
This has been unblocked by #26042, and it has been updated ready for another review.

Lalit3716 commented:
Okay so I have resolved all the conflicts, it is ready for a look.

PIG208 commented:
@zulipbot add "integration review"

PIG208 commented:
I think the PR is now in a good state for an integration review. We might not be able to capture all inaccuracies at this point, but we should be able to refine them as more modules are migrated to TypeScript. Eventually, we want to enforce stricter checks with ESlint and do another sweep. In the future, commits like `"(squash) types: Extend Message and RawMessage types."` can be combined with `"(squash) people: Type Message-Related functions."`. Because during the review it would be easier to refer to the types when they are defined and used in the same commit.

PIG208 commented:
@zulipbot add "integration review"
@zulipbot remove "maintainer review" "mentor review"

timabbott commented:
Squashed and merged, thanks so much for all the work on this @Lalit3716 and @PIG208!!
